src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/update/action/OnRevertSubFlowAllocatedAction.java

```diff
@@ -0,0 +1,99 @@
+/* Copyright 2021 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.flowhs.fsm.yflow.update.action;
+
+import static java.lang.String.format;
+
+import org.openkilda.messaging.command.yflow.SubFlowDto;
+import org.openkilda.messaging.command.yflow.SubFlowSharedEndpointEncapsulation;
+import org.openkilda.messaging.error.ErrorType;
+import org.openkilda.model.Flow;
+import org.openkilda.model.FlowEndpoint;
+import org.openkilda.model.YFlow;
+import org.openkilda.model.YSubFlow;
+import org.openkilda.persistence.PersistenceManager;
+import org.openkilda.persistence.repositories.FlowRepository;
+import org.openkilda.persistence.repositories.RepositoryFactory;
+import org.openkilda.persistence.repositories.YFlowRepository;
+import org.openkilda.persistence.tx.TransactionManager;
+import org.openkilda.wfm.topology.flowhs.exception.FlowProcessingException;
+import org.openkilda.wfm.topology.flowhs.fsm.yflow.update.YFlowUpdateContext;
+import org.openkilda.wfm.topology.flowhs.fsm.yflow.update.YFlowUpdateFsm;
+import org.openkilda.wfm.topology.flowhs.fsm.yflow.update.YFlowUpdateFsm.Event;
+import org.openkilda.wfm.topology.flowhs.fsm.yflow.update.YFlowUpdateFsm.State;
+
+import lombok.extern.slf4j.Slf4j;
+import org.squirrelframework.foundation.fsm.AnonymousAction;
+
+@Slf4j
+public class OnRevertSubFlowAllocatedAction extends AnonymousAction<YFlowUpdateFsm, State, Event, YFlowUpdateContext> {
+    private final TransactionManager transactionManager;
+    private final YFlowRepository yFlowRepository;
+    protected final FlowRepository flowRepository;
+
+    public OnRevertSubFlowAllocatedAction(PersistenceManager persistenceManager) {
+        this.transactionManager = persistenceManager.getTransactionManager();
+        RepositoryFactory repositoryFactory = persistenceManager.getRepositoryFactory();
+        this.yFlowRepository = repositoryFactory.createYFlowRepository();
+        this.flowRepository = repositoryFactory.createFlowRepository();
+    }
+
+    @Override
+    public void execute(State from, State to, Event event, YFlowUpdateContext context, YFlowUpdateFsm stateMachine) {
+        String subFlowId = context.getSubFlowId();
+        if (!stateMachine.isUpdatingSubFlow(subFlowId)) {
```

Comment:
Is it possible to get here outdated response (for flow not listed in "updating sub flows" list in case of global/regular retry?

Reply:
No, it's not possible. Until we receive a completion message (success/failure) of the previous request, we will not do anything else. And if we received a message about the completion, then we will definitely not notify about the allocation of resources.